src/Program2.cs

```diff
@@ -0,0 +1,535 @@
+                case 4:
+                    // yield return 43;
+                    yieldReturn(43, state: 5);
+                    return;
+                case 5:
+                    // yield return 44;
+                    yieldReturn(44, state: 6);
+                    return;
+                case 6:
+                    // await Done();
+                    awaitExpr(state: 7, fast: true);
+                    goto case 7;
+                case 7:
+                    // yield return 45;
+                    yieldReturn(45, state: 9);
+                    return;
+                case 9:
+                    // yield return 46;
+                    yieldReturn(46, state: 10);
+                    return;
+                case 10:
+                    end();
+                    return;
+            }
+
+            // yield handshake:
+            void yieldReturn(int value, int state)
+            {
+                _current = value;
+
+                int previousState = State;
+                State = state;
+
+                if (this._promiseIsActive)
+                {
+                    // reaching a `yield` following an `await`
+                    _valueOrEndPromise.SetResult(true);
+                }
+                else if (previousState == -1)
+                {
+                    // TODO: what if we came from a short-circuited await?
+
+                    // reaching a `yield` from start
+                    // If we came from the start, we'll pretend that the state machine ran.
+                    // This way, the value will be properly yielded in the following TryGetNext
+                    _promiseIsActive = true;
+                    _valueOrEndPromise.SetResult(true);
+                }
+            }
+
+            // await handshake:
+            void awaitExpr(int state, bool fast)
+            {
+                Task task = fast ? Task.CompletedTask : Task.Delay(100);
+                this.Awaiter = task.GetAwaiter();
+                State = state;
+                if (!task.IsCompleted)
+                {
+                    if (!this._promiseIsActive)
+                    {
+                        _promiseIsActive = true;
+                        _valueOrEndPromise.Reset();
+                    }
+                    var self = this;
+                    Builder.AwaitOnCompleted(ref Awaiter, ref self);
+                }
+            }
+
+            void end()
+            {
+                State = -2;
+                if (_promiseIsActive)
+                {
+                    _valueOrEndPromise.SetResult(false);
+                }
+            }
+        }
+
+        // PROTOTYPE(async-streams): update interface definition and async-foreach should recognize task-like in pattern
+        // Contract: WaitForNextAsync will either return a false (no value left) or a true (found a value) with the promise set (to signal an unreturned value is stored).
+        public ValueTask<bool> WaitForNextAsync()
```

Comment:
Should `WaitForNextAsync`, `TryGetNext` and the interface/implementation for the promise be pulled into a base class?